src/vmspawn/vmspawn-scope.c

```diff
@@ -0,0 +1,383 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <stdio.h>
+
+#include "sd-bus.h"
+
+#include "bus-error.h"
+#include "bus-locator.h"
+#include "bus-unit-util.h"
+#include "bus-util.h"
+#include "bus-wait-for-jobs.h"
+#include "macro-fundamental.h"
+#include "macro.h"
+#include "process-util.h"
+#include "random-util.h"
+#include "strv.h"
+#include "unit-def.h"
+#include "unit-name.h"
+#include "vmspawn-scope.h"
+
+int start_transient_scope(sd_bus *bus, const char *machine_name, char **ret_scope) {
+        _cleanup_(bus_wait_for_jobs_freep) BusWaitForJobs *w = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        _cleanup_free_ char *unmangled = NULL, *scope = NULL, *description = NULL;
+        const char *object;
+        int r;
+
+        /* Creates a transient scope unit which tracks the lifetime of the current process */
+
+        r = bus_wait_for_jobs_new(bus, &w);
+        if (r < 0)
+                return log_error_errno(r, "Could not watch job: %m");
+
+        if (asprintf(&unmangled, "sd-vmspawn-%"PRIu32"-%s", random_u32(), machine_name) < 0)
+                return log_oom();
+
+        r = unit_name_mangle_with_suffix(unmangled, "as scope name", 0, ".scope", &scope);
+        if (r < 0)
+                return r;
+
+        description = strjoin("Virtual Machine ", machine_name);
+        if (!description)
+                return log_oom();
+
+        r = bus_call_method(bus, bus_systemd_mgr, "StartTransientUnit", &error, &reply,
+                            "ssa(sv)a(sa(sv))",
+                            /* ss - name, mode */
+                            scope, "fail",
+                            /* a(sv) - properties */
+                            3,
+                            "Description", "s",  description,
+                            "AddRef",      "b",  1,
+                            "PIDs",        "au", 1, getpid_cached(),
```

Comment:
please use pidfds for this now. see how newer nspawn does this.

Reply:
I tried this originally but my systemd version didn't support it so I assumed I had it wrong. I've now added a similar try with pidfd then without on failure that nspawn does.
Fixed in 37e75950ff.